implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex

```diff
@@ -1,380 +1,611 @@
-          | {:error, {:init_timeout, {pid :: pid(), address :: Ockam.Address.t()}}}
-          | {:error, reason :: any()}
-  def create(options, timeout \\ 30_000) when is_list(options) do
-    ## TODO: why secure channel is not a worker?
-    address_prefix = Keyword.get(options, :address_prefix, "")
-
-    options =
-      Keyword.put_new_lazy(options, :address, fn ->
-        Node.get_random_unregistered_address(address_prefix)
-      end)
-
-    case Node.start_supervised(__MODULE__, options) do
-      {:ok, pid, address} ->
-        with {:ok, _pid, address} <- wait_for_channel(pid, address, timeout) do
-          {:ok, address}
-        end
-
-      error ->
-        error
+  # Deprecated, use start_link_channel/2
+  @spec create_channel([initiator_opt()], non_neg_integer) ::
+          {:ok, Ockam.Address.t()} | {:error, term()}
+  def create_channel(opts, handshake_timeout) do
+    ref = make_ref()
+
+    opts =
+      Keyword.merge(opts,
+        handshake_timeout: handshake_timeout,
+        role: :initiator,
+        waiter: {self(), ref},
+        restart_type: :temporary
+      )
+
+    {:ok, addr} = create(opts)
+
+    receive do
+      {:connected, ^ref} -> {:ok, addr}
+    after
+      handshake_timeout ->
+        Ockam.Node.stop(addr)
+        {:error, :key_exchange_timeout}
     end
   end
 
-  def wait_for_channel(pid, address, timeout, interval \\ 100)
-
-  def wait_for_channel(pid, address, timeout, _interval)
-      when is_integer(timeout) and timeout < 0 do
-    {:error, {:init_channel, :timeout, {pid, address}}}
+  def get_remote_identity(worker) do
+    Ockam.Worker.call(worker, :get_remote_identity)
   end
 
-  def wait_for_channel(pid, address, timeout, interval) when is_integer(timeout) do
-    case established?(address) do
-      false ->
-        :timer.sleep(interval)
-        wait_for_channel(pid, address, timeout - interval, interval)
-
-      true ->
-        {:ok, pid, address}
-    end
+  def get_remote_identity_id(worker) do
+    Ockam.Worker.call(worker, :get_remote_identity_id)
   end
 
-  def wait_for_channel(pid, address, :infinity, interval) do
-    case established?(address) do
-      false ->
-        :timer.sleep(interval)
-        wait_for_channel(pid, address, :infinity, interval)
+  @doc """
+  Stop secure channel and it's remote endpoint
+  """
+  def disconnect(worker) do
+    # TODO: a better solution is needed, this works
+    # in a best-effort manner as long as disconnect() is being called
+    Ockam.Worker.call(worker, :disconnect)
+  end
 
-      true ->
-        {:ok, pid, address}
-    end
+  def role(worker) do
+    Ockam.Worker.call(worker, :role)
   end
 
-  @doc false
-  def start_link(options) when is_list(options) do
-    address = Keyword.get(options, :address, Node.get_random_unregistered_address())
-    options = Keyword.put(options, :address, address)
+  def established?(worker) do
+    Ockam.Worker.call(worker, :established?)
+  end
```

Comment:
Optional, if we're not going to supply docstrings for some of these, and they're just one-to-one wrappers of `call`, I'd consider metaprogramming them for brevity:

```
for msg <- ~w(disconnect established? get_remote_identity get_remote_identity_id role)a do # and so on
  def unquote(msg)(worker), do: Ockam.Worker.call(worker, unquote(msg))
end

Reply:
will pass on that :),   having type specs and docstrings will be helpful, we should start adding those at least on this kind of more user-facing places.